File: CHUONG_7/7.09_MultiThread/app/src/main/java/com/example/myapplication/adapter/SongListAdapter.java
package com.example.myapplication.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.example.myapplication.R;
import com.example.myapplication.model.Song;

import java.util.List;

public class SongListAdapter extends RecyclerView.Adapter<SongListAdapter.ViewHolder> {
    private final List<Song> songs;
    private final Context context;

    public SongListAdapter(Context context, List<Song> songs) {
        this.songs = songs;
        this.context = context;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.song_item, parent, false);
        return new ViewHolder(context, view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bind(songs.get(position));
    }

    @Override
    public int getItemCount() {
        return songs.size();
    }

    public void updateSongs(List<Song> songs) {
        this.songs.clear();
        this.songs.addAll(songs);
        notifyDataSetChanged();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        private Context context;
        private TextView textSongTitle;
        private TextView textSongArtist;
        private ImageView imageSong;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
        }

        public ViewHolder (Context context, @NonNull View view){
            this(view);
            this.context = context;
            textSongArtist = view.findViewById(R.id.text_song_artist);
            textSongTitle = view.findViewById(R.id.text_song_title);
            imageSong = view.findViewById(R.id.img_song);
        }

        public void bind(Song song) {
            this.textSongArtist.setText(song.getArtist());
            this.textSongTitle.setText(song.getTitle());
            Glide.with(context).load(song.getImage()).error(R.drawable.ic_album)
                    .into(imageSong);
        }
    }
}

File: CHUONG_7/7.08_Coroutine/app/src/main/java/com/example/coroutine/adapter/SongListAdapter.kt
package com.example.coroutine.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.example.coroutine.R
import com.example.coroutine.model.Song
import com.google.android.material.snackbar.Snackbar

class SongListAdapter (
    private val context: Context,
    private val songs: MutableList<Song>
): RecyclerView.Adapter<SongListAdapter.ViewHolder>() {
    class ViewHolder(
        private val context: Context,
        private val view: View
    ): RecyclerView.ViewHolder(view) {
        private val textArtist = view.findViewById<TextView>(R.id.text_song_artist)
        private val textTitle = view.findViewById<TextView>(R.id.text_song_title)
        private val imageSong = view.findViewById<ImageView>(R.id.img_album)

        fun bind(song: Song) {
            textTitle.text = song.title
            textArtist.text = song.artist
            Glide.with(context)
                .load(song.image)
                .error(R.drawable.ic_album)
                .into(imageSong)
            view.setOnClickListener {
                Snackbar.make(
                    view,
                    "Bạn vừa chọn: ${song.title} - ${song.artist}",
                    Snackbar.LENGTH_SHORT
                ).show()
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val adapterView = LayoutInflater.from(context).inflate(R.layout.song_item, parent, false)
        return ViewHolder(context, adapterView)
    }

    override fun getItemCount(): Int {
        return songs.size
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.bind(songs[position])
    }

    fun updateSongs(songs: List<Song>){
        this.songs.clear()
        this.songs.addAll(songs)
        notifyDataSetChanged()
    }
}